server/routerlicious/packages/lambdas/src/nexus/connect.ts

```diff
@@ -0,0 +1,604 @@
+			getLumberBaseProperties(connection.documentId, connection.tenantId),
+		);
+	}
+
+	connectDocumentOrdererConnectionMetric.success("Successfully established orderer connection");
+
+	return composeConnectedMessage(
+		connection.maxMessageSize,
+		"write",
+		connection.serviceConfiguration.blockSize,
+		connection.serviceConfiguration.maxMessageSize,
+		clientId,
+		claims,
+		version,
+		clients,
+	);
+}
+
+function trackSocket(
+	socket: IWebSocket,
+	tenantId: string,
+	documentId: string,
+	claims: ITokenClaims,
+	{ socketTracker }: INexusLambdaSettings,
+) {
+	// Track socket and tokens for this connection
+	if (socketTracker && claims.jti) {
+		socketTracker.addSocketForToken(
+			createCompositeTokenId(tenantId, documentId, claims.jti),
+			socket,
+		);
+	}
+}
+
+function checkThrottle(tenantId: string, { throttlers, logger }: INexusLambdaSettings): void {
+	const throttleErrorPerCluster = checkThrottleAndUsage(
+		throttlers.connectionsPerCluster,
+		getSocketConnectThrottleId("connectDoc"),
+		tenantId,
+		logger,
+	);
+	if (throttleErrorPerCluster) {
+		// eslint-disable-next-line @typescript-eslint/no-throw-literal
+		throw throttleErrorPerCluster;
+	}
+	const throttleErrorPerTenant = checkThrottleAndUsage(
+		throttlers.connectionsPerTenant,
+		getSocketConnectThrottleId(tenantId),
+		tenantId,
+		logger,
+	);
+	if (throttleErrorPerTenant) {
+		// eslint-disable-next-line @typescript-eslint/no-throw-literal
+		throw throttleErrorPerTenant;
+	}
+}
+
+async function checkToken(
+	token: string | null,
+	tenantId: string,
+	documentId: string,
+	{ tenantManager, revokedTokenChecker, logger }: INexusLambdaSettings,
+): Promise<ITokenClaims> {
+	if (!token) {
+		throw new NetworkError(403, "Must provide an authorization token");
+	}
+	const claims = validateTokenClaims(token, documentId, tenantId);
+	try {
+		if (revokedTokenChecker && claims.jti) {
+			const isTokenRevoked: boolean = await revokedTokenChecker.isTokenRevoked(
+				claims.tenantId,
+				claims.documentId,
+				claims.jti,
+			);
+			if (isTokenRevoked) {
+				throw new TokenRevokedError(
+					403,
+					"Permission denied. Token has been revoked",
+					false /* canRetry */,
+					true /* isFatal */,
+				);
+			}
+		}
+		await tenantManager.verifyToken(claims.tenantId, token);
+		return claims;
+	} catch (error: any) {
+		if (isNetworkError(error)) {
+			throw error;
+		}
+		// We don't understand the error, so it is likely an internal service error.
+		const errMsg = `Could not verify connect document token. Error: ${safeStringify(
+			error,
+			undefined,
+			2,
+		)}`;
+		throw handleServerErrorAndConvertToNetworkError(
+			logger,
+			errMsg,
+			claims.documentId,
+			claims.tenantId,
+			error,
+		);
+	}
+}
+
+async function joinRoomAndSubscribeToChannel(
+	socket: IWebSocket,
+	tenantId: string,
+	documentId: string,
+	{ logger }: INexusLambdaSettings,
+): Promise<[string, IRoom]> {
+	const clientId = generateClientId();
```

Comment:
I think for each of this stage, I added a state to record where the connection state stoppped/failed originally. Do we still want to do this?

Reply:
Those are down in `connectDocument` method where these are called: https://github.com/microsoft/FluidFramework/pull/19906/files/90836d9394f3dac1e97a7ac4dd69c772dac6bd42#diff-043c8ac2e224ef032226c7191a6e23f95b2139f65a140236dd2872667be1a115R488